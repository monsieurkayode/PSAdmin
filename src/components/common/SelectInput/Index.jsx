import React from 'react';
import {
  string, func, arrayOf, shape, bool
} from 'prop-types';
import Loader from '../Loader/Index';

/**
 * @function SelectInput
 *
 * @param {object} props
 *
 * @returns {JSX}
 */
const SelectInput = ({
  name,
  label,
  defaultOption,
  onChange,
  value,
  error,
  options,
  loading,
  onFocus
}) => (
  <div className="form-group">
    <label htmlFor={name}>{label}</label>
    <div className="field custom-field">
      <select
        name={name}
        className="form-control"
        onChange={onChange}
        onFocus={onFocus}
        value={value}
      >
        <option value="">{defaultOption}</option>
        {options.map(option => (
          <option key={option.value} value={option.value}>{option.text}</option>
        ))}
      </select>
      {loading && <Loader size={20} />}
    </div>
    {error && (
      <div
        style={{ maxWidth: '93%', padding: '5px' }}
        className="alert alert-danger"
      >
        {error}
      </div>
    )}
  </div>
);

SelectInput.defaultProps = {
  value: '',
  error: ''
};

SelectInput.propTypes = {
  name: string.isRequired,
  label: string.isRequired,
  defaultOption: string.isRequired,
  onChange: func.isRequired,
  options: arrayOf(shape({})).isRequired,
  value: string,
  error: string,
  loading: bool.isRequired,
  onFocus: func.isRequired
};

export default SelectInput;
