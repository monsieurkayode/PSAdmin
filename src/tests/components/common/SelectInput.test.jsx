import React from 'react';
import { mount } from 'enzyme';
import SelectInput from '../../../components/common/SelectInput/Index';

let wrapper;

const setup = (loading, error) => {
  const props = {
    name: '',
    label: '',
    defaultOption: '',
    value: '',
    onChange: jest.fn(),
    onFocus: jest.fn(),
    error,
    options: [{ value: '', text: '' }],
    loading
  };

  return mount(<SelectInput {...props} />);
};

describe('SelectInput Component', () => {
  it('renders correctly', () => {
    wrapper = setup(false, '');
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a label and select element', () => {
    expect(wrapper.find('label').length).toEqual(1);
    expect(wrapper.find('select').length).toEqual(1);
  });

  it('renders a loader component when making async requests', () => {
    wrapper = setup(true, '');
    expect(wrapper.find('Loader').length).toEqual(1);
  });

  it('does not render loader component when not making async requests', () => {
    wrapper = setup(false, '');
    expect(wrapper.find('Loader').length).toEqual(0);
  });

  it('renders an error div when there are errors', () => {
    wrapper = setup(false, 'Select an author');
    expect(wrapper.find('.alert').length).toEqual(1);
  });
});
