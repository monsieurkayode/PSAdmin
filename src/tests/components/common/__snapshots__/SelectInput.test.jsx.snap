// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput Component renders correctly 1`] = `
<SelectInput
  defaultOption=""
  error=""
  label=""
  loading={false}
  name=""
  onChange={[MockFunction]}
  onFocus={[MockFunction]}
  options={
    Array [
      Object {
        "text": "",
        "value": "",
      },
    ]
  }
  value=""
>
  <div
    className="form-group"
  >
    <label
      htmlFor=""
    />
    <div
      className="field custom-field"
    >
      <select
        className="form-control"
        name=""
        onChange={[MockFunction]}
        onFocus={[MockFunction]}
        value=""
      >
        <option
          value=""
        />
        <option
          key=""
          value=""
        />
      </select>
    </div>
  </div>
</SelectInput>
`;
